### ISSUE TITLE
CultureNotFoundException in Net6 but works in Net5 on same system

### ISSUE THREAD

**Issue by igandrews** [dotnet/runtime#62929]
### Description

We create some cultures to do string comparisons for our library. That requires access to alternate sort orders. Prior to .net 5 we were using new CultureInfo(0x30404). However with the change to ICU in .net 5 the culture that produced did not provide the same sort/compare results. We found that new CultureInfo("zh-TW-u-co-zhuyin") produced better results and so started using that in .net 5. However the same code on the same system but in .net 6 throws a CultureNotFoundException with a message of:
```
Culture is not supported. (Parameter 'name')
zh-TW-u-co-zhuyin is an invalid culture identifier.
```
In .net 5 the culture is found and created and the CultureInfo has an EnglishName 'Chinese (Taiwan, Sort Order=Zhuyin Sort Order)' and .  Prior to .net 5 we try to use the lcid specified above but if it did hit this path it just creates an unknown culture (the EnglishName reads "Unknown Locale (zh-TW-u-co-zhuyin)"). So why is .net 5 able to load it and .net 6 is not?

### Reproduction Steps

var c = new CultureInfo("zh-TW-u-co-zhuyin");

### Expected behavior

This should construct the culture (with the correct CompareInfo) as it does in .net 5

### Actual behavior

Culture is not supported. (Parameter 'name')
zh-TW-u-co-zhuyin is an invalid culture identifier.

### Regression?

Yes this is a regression. In .Net 5 the culture is created. Prior to .net 5 an unknown culture with the default compareinfo is created but then we didn't use this then anyway - we used the 0x30404 as I mentioned above.

### Configuration

.NET 6 (6.00.21.52210)
Windows 10.0.19044

### Other information

The exception is not thrown when the unicode extension tag is excluded but then it also doesn't actually sort correctly. In .net 5 sorting with the culture created as mentioned above and then comparing like the following would return -1:
`c.CompareInfo.Compare("巴斯克語", "土耳其語")`
But if I remove the -u unicode extension the culture created (whose EnglishName is then 'Chinese (Taiwan, CO_ZHUYIN)' in .net 5 and .net 6) would return 1 for the above. Please don't suggest "zh-TW_zhuyin" either. That too will construct a culture but not sort as desired.

**Comment by tarekgh** [comment]
@igandrews thanks for this report too. The behavior change in .NET 6.0 is because the culture name needs to be supported by Windows too as we pick some properties from Windows even when we are using ICU. 

You may try using the name `zh-TW_u-co-zhuyin` and let's know if you will get the desired behavior you want.

**Comment by igandrews** [comment]
While it does not blow up the result is a bit different even if I use that name in .net 5. So for clarification:

Code used to test:
```c#
Action<CultureInfo> test = (c) => Console.WriteLine($"{c.Name} {c.EnglishName} {c.CompareInfo.Compare("巴斯克語", "土耳其語")} {c.CompareInfo.Compare("葡萄牙語", "丹麥語")}");

test(new CultureInfo("zh-TW-u-co-zhuyin"));
test(new CultureInfo("zh-TW_u-co-zhuyin"));
```

.Net 5 results:
```
zh-TW-u-co-zhuyin Chinese (Taiwan, Sort Order=Zhuyin Sort Order) -1 -1
zh-TW-u-co-zhuyin Chinese (Taiwan, Sort Order=Zhuyin Sort Order) -1 -1
```

.Net 6 results (assuming you comment out the 1st test with "zh-TW-u-co...":
```
zh-TW_u-co-zhuyin Chinese (Taiwan, Sort Order=u-co-zhuyin) -1 1
```

So while I'm not particularly concerned about the name the fact that they appear different seems to indicate .net 6 is handling it differently. More importantly the result of the 2nd string comparison is incorrect.

BTW it seems odd to me that MS changed from NLS to ICU in .net 5 (a breaking change for dev's who may have never intended to run their code on other platforms) under the statement that this was done to make things consistent across the platforms only to then make changes in .net 6 that tie behavior back to the platform. I mean would the name `zh-TW-u-co-zhuyin` have thrown on non-windows platforms in .net 6? Would using `zh-TW_u-co-zhuyin` work on non-windows platforms?

**Comment by tarekgh** [comment]
@igandrews thanks for your reply. 

> BTW it seems odd to me that MS changed from NLS to ICU in .net 5 (a breaking change for dev's who may have never intended to run their code on other platforms) under the statement that this was done to make things consistent across the platforms only to then make changes in .net 6 that tie behavior back to the platform. 

For devs who want to run on Windows only and want to continue getting the old behavior (i.e. NLS behavior), they can still do that using the config switch mentioned in the [doc](https://docs.microsoft.com/en-us/dotnet/core/extensions/globalization-icu#use-nls-instead-of-icu). 
You have mentioned we have a change in .NET 6.0 that ties behavior to the platform. This is not fully true. We still have consistency across all platforms. The only special case is the locale names that cannot work on Windows we disallow that because it can fail later (for example if need to pick the user overrides properties from the system). Such locale names that not supported by Windows consider rare and 99.9% of the other cases already supported by Windows. 

> I mean would the name zh-TW-u-co-zhuyin have thrown on non-windows platforms in .net 6? Would using zh-TW_u-co-zhuyin work on non-windows platforms?

Actually, `zh-TW_u-co-zhuyin` is failing even in .NET 5.0 when trying to compare strings with it. So, it just happened that this culture worked on Windows and was not intentional to have such discrepancy. 

To summarize the issue here:

- `zh-TW_u-co-zhuyin` should be equivalent `zh-TW_zhuyin` but I have looked, and I am seeing some bug which didn't allow this mapping. In ICU `zh-TW_u-co-zhuyin` should be equivalent to `zh-TW@collation=zhuyin`. 
- The real problem is not about allowing `zh-TW_u-co-zhuyin`, but it is about making it to behave as `zh-TW_zhuyin`.

Now, my question to you, how is this blocking you? Would it be ok if we fix it in the next release (.NET 7.0)?

**Comment by tarekgh** [comment]
The fix of this issue will be in .NET 7.0. Let me know if you have any questions.
